setup.py

```diff
@@ -66,9 +68,91 @@
 from setuptools.command.build_ext import build_ext
 from setuptools.command.build_py import build_py
 
+# Set up logging
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s [ExecuTorch] %(levelname)s: %(message)s"
+)
+logger = logging.getLogger()
+
 # For information on setuptools Command subclassing see
 # https://setuptools.pypa.io/en/latest/userguide/extension.html
 
+################################################################################
+# Git submodules
+################################################################################
+# The following submodules are required to be able to build ExecuTorch. If any of
+# these folders are missing or missing CMakeLists.txt, we will run
+# `git submodule update` to try to fix it. If the command fails, we will raise an
+# error.
+# An alternative to this would be to run `git submodule status` and run
+# `git submodule update` if there's any local changes. However this is a bit
+# too restrictive for users who modifies and tests the dependencies locally.
+
+# keep sorted
+REQUIRED_SUBMODULES = {
+    "ao": "LICENSE",  # No CMakeLists.txt, choose a sort of stable file to check.
+    "cpuinfo": "CMakeLists.txt",
+    "eigen": "CMakeLists.txt",
+    "flatbuffers": "CMakeLists.txt",
+    "FP16": "CMakeLists.txt",
+    "FXdiv": "CMakeLists.txt",
+    "gflags": "CMakeLists.txt",
+    "prelude": "BUCK",
+    "pthreadpool": "CMakeLists.txt",
+    "pybind11": "CMakeLists.txt",
+    "XNNPACK": "CMakeLists.txt",
+}
+
+
+def get_required_submodule_paths():
+    gitmodules_path = os.path.join(os.getcwd(), ".gitmodules")
+
+    if not os.path.isfile(gitmodules_path):
+        logger.error(".gitmodules file not found.")
+        exit(1)
+
+    with open(gitmodules_path, "r") as file:
+        lines = file.readlines()
+
+    # Extract paths of required submodules
+    required_paths = {}
+    for line in lines:
+        if line.strip().startswith("path ="):
+            path = line.split("=")[1].strip()
+            for submodule, file_name in REQUIRED_SUBMODULES.items():
+                if submodule in path:
+                    required_paths[path] = file_name
+    return required_paths
+
+
+def check_and_update_submodules():
+    def check_folder(folder: str, file: str) -> bool:
+        return os.path.isdir(folder) and os.path.isfile(os.path.join(folder, file))
+
+    # Check if the directories exist for each required submodule
+    missing_submodules = {}
+    for path, file in get_required_submodule_paths().items():
+        if not check_folder(path, file):
+            missing_submodules[path] = file
+
+    # If any required submodule directories are missing, update them
+    if missing_submodules:
+        logger.warning("Some required submodules are missing. Updating submodules...")
+        try:
+            subprocess.check_call(["git", "submodule", "sync"])
```

Comment:
I don't like doing this in setup.py because it adds a new assumption that the code lives in git. Until now, we could build from a tarball of the entire tree outside of git. If we were going to do it anywhere I'd prefer doing it in install_executorch or install_requirements: setup.py should not be modifying its environment, only building it.